src/LightningDB/LightningCursor.cs

```diff
@@ -265,36 +289,114 @@ private bool GetMultiple(CursorOperation operation)
+        ///     CursorPutOptions.ReserveSpace - reserve space for data of the given size, but don't copy the given data. Instead, return a pointer to the reserved space, which the caller can fill in later. This saves an extra memcpy if the data is being generated later.
+        ///     CursorPutOptions.AppendData - append the given key/data pair to the end of the database. No key comparisons are performed. This option allows fast bulk loading when keys are already known to be in the correct order. Loading unsorted keys with this flag will cause data corruption.
+        ///     CursorPutOptions.AppendDuplicateData - as above, but for sorted dup data.
+        /// </param>
+        public unsafe void Put(ReadOnlySpan<byte> key, ReadOnlySpan<byte> value, CursorPutOptions options)
         {
-            var mdbKey = new MDBValue(new Span<byte>(key));
-            var mdbValue = new MDBValue(new Span<byte>(values.SelectMany(x => x).ToArray()), GetSize(values));
-            var mdbCount = new MDBValue
+            fixed (byte* keyPtr = key)
+            fixed (byte* valPtr = value)
             {
-                size = (IntPtr) values.Length
-            };
-            mdb_cursor_put(_handle, ref mdbKey, new []{ mdbValue, mdbCount }, CursorPutOptions.MultipleData);
+                var mdbKey = new MDBValue(key.Length, keyPtr);
+                var mdbValue = new MDBValue(value.Length, valPtr);
+
+                mdb_cursor_put(_handle, mdbKey, mdbValue, options);
+            }
         }
-        
-        private int GetSize(byte[][] values)
+
+        /// <summary>
+        /// Store by cursor.
+        /// This function stores key/data pairs into the database. 
+        /// If the function fails for any reason, the state of the cursor will be unchanged. 
+        /// If the function succeeds and an item is inserted into the database, the cursor is always positioned to refer to the newly inserted item.
+        /// </summary>
+        /// <param name="key">The key operated on.</param>
+        /// <param name="values">The data items operated on.</param>
+        public unsafe void PutMultiple(byte[] key, byte[][] values)
         {
-            if (values.Length == 0 || values[0] == null || values[0].Length == 0)
-                return 0;
+            const int StackAllocateLimit = 256;//I just made up a number, this can be much more agressive -arc
+
+            int overallLength = values.Sum(arr => arr.Length);//probably allocates but boy is it handy...
+
+
+            //the idea here is to gain some perf by stackallocating the buffer to 
+            //hold the contiguous keys
+            if (overallLength < StackAllocateLimit)
+            {
+                Span<byte> contiguousValues = stackalloc byte[overallLength];
 
-            return values[0].Length;
+                InnerPutMultiple(contiguousValues);
+            }
+            else
+            {
+                fixed (byte* contiguousValuesPtr = new byte[overallLength])
```

Comment:
Not sure if new buffers are created elsewhere, or indeed how often you'd expect this unoptimised path to be hit, but maybe it could use pooled memory instead of creating a new byte array, e.g. using `ArrayPool` from `System.Buffers`?

(just a random user happy to see Span `support`, having a quick browse through the PR 😄)

Reply:
The previous implementation allocated a lot more also, so this should be a net improvement either way.

As for pooling vs allocs, I've actually been kinda frustrated at how good the .NET GC has become 😅.  I haven't benchmarked this method specifically, but it is probably still improvable, the question is by how much. 

If this path is important to your use case I can create an issue for it to investigate.

Reply:
I haven't actually benchmarked (or even tried the new Spanified APIs yet), I was just looked at the PR and happened to notice this, and it's similar to a pattern I use a lot (fast path using `stackalloc` for small buffers, slow path using `ArrayPool` for larger buffers). But for sure as-is it's a hige net benefit 😃